Add a platform config to disable the MPU

c1728p9 opened the pull request:
### Description

Add the platform config DISABLE_MPU to allow the MPU to be disabled from an application.

This PR is identical to #9030 but has an additional commit to add this config.

### Pull request type

    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

c1728p9 committed:
Add a platform config to disable the MPU

Add the platform config DISABLE_MPU to allow the MPU to be disabled
from an application.

ciarmcom commented:
@c1728p9, thank you for your changes.
@ARMmbed/mbed-os-hal @ARMmbed/mbed-os-core @ARMmbed/mbed-os-test @ARMmbed/mbed-os-maintainers please review.

deepikabhavnani commented on platform/mbed_mpu_mgmt.c:
```diff
@@ -17,19 +17,19 @@
 static uint16_t mem_xn_lock;
 static uint16_t mem_wn_lock;
 
 void mbed_mpu_manager_lock_ram_execution()
```
Functions are enabled only when `	#if DEVICE_MPU && !MBED_CONF_PLATFORM_DISABLE_MPU` is true, shouldn;t it be same for ScopedRamExecutionLock.h and ScopedRomWriteLock.h

Also since we are trying to reduce size, should be use direct lock calls in FlashIAP, instead of `Scoped` class layer?

c1728p9 replied:
@deepikabhavnani initially I wrapped ScopedRamExecutionLock and ScopedRomWriteLock in `#if DEVICE_MPU && !MBED_CONF_PLATFORM_DISABLE_MPU`  but I found it didn't actually save any space, as the compiler was already removing it. I was using a GCC_ARM development build.

c1728p9 replied:
Debug builds would be forced to keep this around though, so that may be a good reason to remove it. Same would go for the RAII class which adds more code at least in debug builds. @kjbracey-arm @donatieng @bulislaw @deepikabhavnani what are your thoughts on this?

deepikabhavnani replied:
If we need RAII class for users we can let it be in code space, but internally in platform and drivers we can use internal functions instead of class.. can check how much space is saved and is worth or not ?

deepikabhavnani replied:
Linker should drop the class and size if not used, in debug builds as well

kjbracey replied:
I'm not too bothered about debug builds - lots of good C++ practice causes bloat if optimisation is fully disabled. You can't start trying to avoid typical teeny inline C++ methods just to "optimise" debug builds - which is something of an oxymoron.

If anything we need more use of scoped locks than manual locks - there's a continuous low-level flow of "add missing unlock" PRs.

cmonr commented:
CI started (in case it's decided to bring in as-is for RC3)

kjbracey commented on platform/mbed_lib.json:
```diff
@@ -128,6 +128,10 @@
             "help": "Setting this to true enables auto-reboot on a fatal error.",
             "value": false
+        },
+        "disable-mpu": {
```
Config options should be positive unless there's a good reason. This could be `enable-mpu`, or maybe `use-mpu` is more to the point.

0xc0170 commented:
> This PR is identical to #9030 but has an additional commit to add this config.

This shall be rebased and some comments addressed above, marking it as "needs: work". 

@kjbracey-arm Do you want to push an update for positive config value? This PR looks minimal and might be considered for rc3?

kjbracey committed:
Rework MPU use option

Make the option positively named, and as it is a platform config
option make sure it only affects platform code.

HAL functions still remain available even if platform is told not
to use them.

kjbracey commented:
Going to rework a little more - as this is a platform option, it should only be affecting platform code. It shouldn't actually be creating dummy HAL calls (and hence requiring custom MPU HALs to follow the config option).

kjbracey commented:
Reworked, and rebased to make the PR clearer. Naming the option "use-mpu" made it clearer how it should be implemented. (ie don't touch the HAL, just don't use it).

0xc0170 commented:
CI was aborted (timeout anyway) - for older commits, will restart now for the latest update

0xc0170 commented:
CI started

adbridge commented:
Looks like device out of space error!

0xc0170 commented:
CI should be fixed (other fixes ongoing)

adbridge commented:
> 
> 
> CI should be fixed (other fixes ongoing)

Other fixes ?

0xc0170 commented:
CI team is checking what caused the space to run out (it was now cleaned)

CI restarted

kjbracey commented on platform/mbed_mpu_mgmt.h:
```diff
@@ -35,6 +35,8 @@ extern "C" {
 
 #define mbed_mpu_manager_init() mbed_mpu_init()
 
+#define #define mbed_mpu_manager_deinit() mbed_mpu_free()
```
Uh-oh.

c1728p9 replied:
At least is isn't #undefined behavior. Just kidding, I pushed a fix.

kjbracey reviewed (approved):
Russ's last commit looks good to me.

cmonr commented:
Ignore https://github.com/ARMmbed/mbed-os/pull/9040#issuecomment-446254844

I accidentally re-enqueued the PR a while back, and forgot to remove it (and force aborted the job above).

deepikabhavnani commented:
@cmonr - I see Commits after the CI run was triggered, last force push 15 mins ago.

cmonr commented:
> @cmonr - I see Commits after the CI run was triggered, last force push 15 mins ago.

@deepikabhavnani Correct. There was a third commit in this PR, but CI was started and [passed ](https://github.com/ARMmbed/mbed-os/pull/9040#issuecomment-446246520) when there were only two commits. We opted to remove the third commit to make this PR green to increase the chance of having RC3 generated before COB UK time.
